components/director/internal/domain/formationconstraint/constraintEngine.go

```diff
@@ -0,0 +1,229 @@
+	c := &ConstraintEngine{
+		constraintSvc:             constraintSvc,
+		tenantSvc:                 tenantSvc,
+		asaSvc:                    asaSvc,
+		formationRepo:             formationRepo,
+		labelRepo:                 labelRepo,
+		operatorInputConstructors: map[OperatorName]OperatorInputConstructor{"IsNotAssignedToAnyFormationOfType": NewIsNotAssignedToAnyFormationOfTypeInput},
+	}
+	c.operators = map[OperatorName]OperatorFunc{"IsNotAssignedToAnyFormationOfType": c.IsNotAssignedToAnyFormationOfType}
+	return c
+}
+
+// EnforceConstraints finds all the applicable constraints based on JoinPointLocation and JoinPointDetails. Checks for each constraint if it is satisfied.
+// If any constraint is not satisfied this information is stored and the engine proceeds with enforcing the next constraint if such exists. In the end if
+// any constraint was not satisfied an error is returned.
+func (e *ConstraintEngine) EnforceConstraints(ctx context.Context, location JoinPointLocation, details JoinPointDetails, formationTemplateID string) error {
+	matchingDetails := details.GetMatchingDetails()
+	log.C(ctx).Infof("Enforcing constraints for target operation %q, constraint type %q, resource type %q and resource subtype %q", location.OperationName, location.ConstraintType, matchingDetails.resourceType, matchingDetails.resourceSubtype)
+
+	constraints, err := e.constraintSvc.ListMatchingConstraints(ctx, formationTemplateID, location, matchingDetails)
+	if err != nil {
+		return errors.Wrapf(err, "While listing matching constraints for target operation %q, constraint type %q, resource type %q and resource subtype %q", location.OperationName, location.ConstraintType, matchingDetails.resourceType, matchingDetails.resourceSubtype)
+	}
+
+	matchedConstraintsNames := make([]string, 0, len(constraints))
+	for _, c := range constraints {
+		matchedConstraintsNames = append(matchedConstraintsNames, c.Name)
+	}
+
+	log.C(ctx).Infof("Matched constraints: %v", matchedConstraintsNames)
+
+	var errs *multierror.Error
+	for _, mc := range constraints {
+		operator, ok := e.operators[OperatorName(mc.Operator)]
+		if !ok {
+			errs = multierror.Append(errs, ConstraintError{
+				ConstraintName: mc.Name,
+				Reason:         fmt.Sprintf("Operator %q not found", mc.Operator),
+			})
+			continue
+		}
+
+		operatorInputConstructor, ok := e.operatorInputConstructors[OperatorName(mc.Operator)]
+		if !ok {
+			errs = multierror.Append(errs, ConstraintError{
+				ConstraintName: mc.Name,
+				Reason:         fmt.Sprintf("Operator input constructor for operator %q not found", mc.Operator),
+			})
+			continue
+		}
+
+		operatorInput := operatorInputConstructor()
+		if err := formation_constraint_input.ParseInputTemplate(mc.InputTemplate, details, operatorInput); err != nil {
```

Comment:
We should add the constraint alongside to the join point detials when rendering the template, otherwise fields like `ConstraintScope` and `OperatorScope` cannot be used in the operator. Or should we demote them to be part of the template directly?

Reply:
The ConstraintScope is not needed by the operator. It is used during constraint matching.

Reply:
For the OperatorScope we could add setter method to the inputs andd set the scope after rendering the template?

Reply:
Removed operator scope as the field is not applicable to most of the constraints. For those which need the field it will be added to the coresponding input template.